common/src/test/java/io/netty/util/HashedWheelTimerTest.java

```diff
@@ -149,9 +160,80 @@ public void run(final Timeout timeout) throws Exception {
         for (int i = 0; i < scheduledTasks; i++) {
             long delay = queue.take();
             assertTrue("Timeout + " + scheduledTasks + " delay " + delay + " must be " + timeout + " < " + maxTimeout,
-                    delay >= timeout && delay < maxTimeout);
+                delay >= timeout && delay < maxTimeout);
         }
 
         timer.stop();
     }
+
+    @Test
+    public void testRejectedExecutionExceptionWhenTooManyTimeoutsAreAddedBackToBack() {
+        HashedWheelTimer timer = new HashedWheelTimer(Executors.defaultThreadFactory(), 100,
+            TimeUnit.MILLISECONDS, 32, true, 2);
+        timer.newTimeout(createNoOpTimerTask(), 1, TimeUnit.SECONDS);
+        timer.newTimeout(createNoOpTimerTask(), 1, TimeUnit.SECONDS);
+        try {
+            timer.newTimeout(createNoOpTimerTask(), 1, TimeUnit.SECONDS);
+            fail("Timer allowed adding 3 timeouts when maxPendingTimeouts was 2");
+        } catch (RejectedExecutionException e) {
+            // Expected
+        } finally {
+            timer.stop();
+        }
+    }
+
+    @Test
+    public void testNewTimeoutShouldStopThrowingRejectedExecutionExceptionWhenExistingTimeoutIsCancelled()
+        throws InterruptedException {
+        final HashedWheelTimer timer = new HashedWheelTimer(Executors.defaultThreadFactory(), 100,
+            TimeUnit.MILLISECONDS, 32, true, 2);
+        timer.newTimeout(createNoOpTimerTask(), 1, TimeUnit.SECONDS);
+        Timeout timeoutToCancel = timer.newTimeout(createNoOpTimerTask(), 1, TimeUnit.SECONDS);
+        timeoutToCancel.cancel();
+        waitUntilNewTimeoutCanBeAdded(200, timer, 1000);
+    }
+
+    @Test
+    public void testNewTimeoutShouldStopThrowingRejectedExecutionExceptionWhenExistingTimeoutIsExecuted()
+        throws InterruptedException {
+        final HashedWheelTimer timer = new HashedWheelTimer(Executors.defaultThreadFactory(), 25,
+            TimeUnit.MILLISECONDS, 4, true, 2);
+        timer.newTimeout(createNoOpTimerTask(), 1, TimeUnit.SECONDS);
+        timer.newTimeout(createNoOpTimerTask(), 90, TimeUnit.MILLISECONDS);
+        waitUntilNewTimeoutCanBeAdded(200, timer, 200);
+        timer.stop();
+    }
+
+    private static void waitUntilNewTimeoutCanBeAdded(final long timeoutMilli,
+        final HashedWheelTimer timer, final long newTimeoutDelay) {
+        waitUntilTrue(timeoutMilli, new Supplier<Boolean>() {
+            @Override
+            public Boolean get() {
+                try {
+                    timer.newTimeout(createNoOpTimerTask(), newTimeoutDelay, TimeUnit.MILLISECONDS);
+                    return true;
+                } catch (RejectedExecutionException e) {
+                    return false;
+                }
+            }
+        });
+    }
+
+    private static void waitUntilTrue(long timeoutMilli, Supplier<Boolean> condition) {
+        long startTime = System.currentTimeMillis();
```

Comment:
`System.currentTimeMillis` is not appropriate for measuring execution time. We should use `nanoTime()` instead